File: app/src/main/java/al/edu/feut/financaime/ExpenseFragment.java
package al.edu.feut.financaime;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.TextInputLayout;
import android.support.v4.app.Fragment;
import android.support.v7.widget.AppCompatEditText;
import android.support.v7.widget.AppCompatSpinner;
import android.support.v7.widget.AppCompatTextView;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.Toast;

import java.util.List;

import al.edu.feut.financaime.model.Budget;
import al.edu.feut.financaime.model.Category;
import al.edu.feut.financaime.model.Database;
import al.edu.feut.financaime.model.Expense;
import al.edu.feut.financaime.util.Utilities;

public class ExpenseFragment extends Fragment {
    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_expense, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        AppCompatSpinner spinner = view.findViewById(R.id.spinner);
        AppCompatTextView budgetHint = view.findViewById(R.id.budget_hint);
        AppCompatEditText inputExpValue = view.findViewById(R.id.input_exp_value);

        TextInputLayout inputExpTextInputLayout = view.findViewById(R.id.input_exp_text_input_layout);

        //Get Data from local db
        Budget budget = new Database(getContext()).budget(Utilities.month(Utilities.month()));
        List<Category> categories = new Database(getContext()).categories();

        //Spinner adapter
        ArrayAdapter<Category> adapter = new ArrayAdapter<>(getContext(), android.R.layout.simple_spinner_dropdown_item, categories);
        spinner.setAdapter(adapter);

        budgetHint.setText(Utilities.format(budget != null ? budget.getBudget() : 0));

        inputExpValue.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void afterTextChanged(Editable editable) {
                inputExpTextInputLayout.setError(null);
            }
        });

        view.findViewById(R.id.btn_add).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (spinner.getSelectedItem().toString().isEmpty() || inputExpValue.getText().toString().isEmpty())
                    inputExpTextInputLayout.setError(getString(R.string.alert));
                else if (budget.getBudget() >= Double.parseDouble(inputExpValue.getText().toString())) {
                    Expense expense = new Expense();
                    expense.setExpenseName(spinner.getSelectedItem().toString());
                    expense.setExpense(Double.parseDouble(inputExpValue.getText().toString()));
                    expense.setDate(Utilities.date());
                    expense.setIdBudget(budget.getId());
                    budget.setExpense(expense);
                    if(new Database(getContext()).insertExpense(expense) != -1) {
                        budget.setBudget(budget.getBudget() - budget.getExpense().getExpense());
                        if(new Database(getContext()).updateBudgetValue(budget) != -1) {
                            Toast.makeText(getActivity(), R.string.success,Toast.LENGTH_SHORT).show();
                            budgetHint.setText(Utilities.format(budget.getBudget()));
                        } else
                            Toast.makeText(getActivity(), R.string.fail,Toast.LENGTH_SHORT).show();
                    }
                }
                else
                    Toast.makeText(getActivity(), R.string.out_of_budget, Toast.LENGTH_SHORT).show();
            }
        });

        view.findViewById(R.id.edit_category).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                getFragmentManager()
                        .beginTransaction()
                        .replace(R.id.host, new ExpenseCategoriesFragment())
                        .addToBackStack("EXPENSE_CATEGORIES_FRAGMENT")
                        .commit();
            }
        });
    }
}

File: app/src/main/java/al/bruno/personal/expense/ExpenseFragment.kt
package al.bruno.personal.expense

import al.bruno.personal.expense.adapter.CustomSpinnerAdapter
import al.bruno.personal.expense.callback.*
import al.bruno.personal.expense.databinding.FragmentExpenseBinding
import android.os.Bundle

import al.bruno.personal.expense.model.Budget
import androidx.fragment.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import al.bruno.personal.expense.util.Utilities.month
import al.bruno.personal.expense.view.model.BudgetViewModel
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders

import al.bruno.personal.expense.databinding.CategoriesSpinnerItemBinding
import al.bruno.personal.expense.model.Categories
import al.bruno.personal.expense.model.Expense
import al.bruno.personal.expense.util.Utilities
import al.bruno.personal.expense.view.model.CategoriesViewModel
import al.bruno.personal.expense.view.model.ExpenseViewModel
import android.util.Log
import android.widget.Toast
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers

class ExpenseFragment : Fragment() {
    private val disposable : CompositeDisposable  = CompositeDisposable()
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val fragmentExpenseBinding : FragmentExpenseBinding = DataBindingUtil.inflate(inflater, R.layout.fragment_expense, container, false)
        ViewModelProviders.of(this).get(BudgetViewModel::class.java).expense(month(month())).observe(this, Observer {
            run {
                fragmentExpenseBinding.budget = it
            }
        })

        disposable.add(ViewModelProviders
                .of(this)
                .get(CategoriesViewModel::class.java)
                .categories()
                .subscribeOn(Schedulers.io())
                .subscribe({
                    fragmentExpenseBinding.spinnerAdapter =
                            CustomSpinnerAdapter(activity!!, R.layout.categories_spinner_item, it, object : BindingData<Categories, CategoriesSpinnerItemBinding> {
                                override fun bindData(t: Categories, vm: CategoriesSpinnerItemBinding) {
                                    vm.categories = t
                                }
                            })
                },{
                    Log.i(ExpenseFragment::class.java.name, it.message)
                }))

        fragmentExpenseBinding.onClick = object : OnClick {
            override fun onClick(){
                fragmentManager!!
                        .beginTransaction()
                        .replace(R.id.host, CategoriesFragment())
                        .addToBackStack("EXPENSE_CATEGORIES_FRAGMENT")
                        .commit()

            }
        }

        fragmentExpenseBinding.onClickListener = object : OnClickListener<Budget> {
            override fun onClick(t: Budget) {
                val expense = Expense()
                expense.expense = t.expenseStr
                expense.amount = t.amount
                expense.date = Utilities.date()
                expense.idBudget = t.id
                disposable.add(ViewModelProviders.of(activity!!)
                        .get(ExpenseViewModel::class.java)
                        .insert(expense)
                        .subscribeOn(Schedulers.io())
                        .subscribe({
                            if(it != -1.toLong()) {
                                t.amount = 0.0
                                Log.i(ExpenseFragment::class.java.name, "Success")
                                //Toast.makeText(activity, R.string.success, Toast.LENGTH_SHORT).show()
                            } else
                                Log.i(ExpenseFragment::class.java.name, "Fail")
                                //Toast.makeText(activity, R.string.fail, Toast.LENGTH_SHORT).show()
                        }, {
                            Log.i(ExpenseFragment::class.java.name, it.message)
                            //Toast.makeText(activity, it.message, Toast.LENGTH_SHORT).show()
                        }))
                Toast.makeText(activity, expense.toString(), Toast.LENGTH_SHORT).show()
            }

        }
        return fragmentExpenseBinding.root
    }

    override fun onStop() {
        super.onStop()
        disposable.clear()
    }
}